File: C++/Medium/Accepted/minimum-absolute-sum-difference_2.cpp
class Solution {
public:
    int minAbsoluteSumDiff(vector<int>& nums1, vector<int>& nums2) {
        
        int n = nums1.size();
        vector<int> sortedNums1(nums1);
        sort(sortedNums1.begin(), sortedNums1.end());
        long sum = 0;
        long maxDiff = 0;
        for (int i = 0; i < n; i++) {
            long diff = abs(nums1[i] - nums2[i]);
            sum += diff;
            int j = lower_bound(sortedNums1.begin(), sortedNums1.end(), nums2[i]) - sortedNums1.begin();
            if (j < n) {
                maxDiff = max(maxDiff, diff - (sortedNums1[j] - nums2[i]));
            }
            if (j > 0) {
                maxDiff = max(maxDiff, diff - (nums2[i] - sortedNums1[j - 1]));
            }
        }
        return (sum - maxDiff) % 1000000007;
    }
};

File: Java/Medium/Accepted/minimum-absolute-sum-difference_2.java
class Solution {
    public int minAbsoluteSumDiff(int[] nums1, int[] nums2) {
        
        int mod = 1000000007;
        int n = nums1.length;
        int[] diff = new int[n];
        for (int i = 0; i < n; i++) {
            diff[i] = Math.abs(nums1[i] - nums2[i]);
        }
        int sum = 0;
        for (int i = 0; i < n; i++) {
            sum = (sum + diff[i]) % mod;
        }
        Arrays.sort(nums1);
        int max = 0;
        for (int i = 0; i < n; i++) {
            int index = binarySearch(nums1, nums2[i]);
            if (index < n) {
                max = Math.max(max, diff[i] - (Math.abs(nums1[index] - nums2[i])));
            }
            if (index > 0) {
                max = Math.max(max, diff[i] - (Math.abs(nums1[index - 1] - nums2[i])));
            }
        }
        return (sum - max + mod) % mod;
    }
    public int binarySearch(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;
        while (low < high) {
            int mid = (high - low) / 2 + low;
            if (nums[mid] < target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        return nums[low] >= target ? low : low + 1;
    }
}